### PR TITLE
feat(menu): Big Navigation Rework for 19.04

### PR DESCRIPTION
**Pull request by victorvassilev**
## Description

Modify the way the navigation works, introduce new api, and new layout structure.

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [x] 19.04.x
- [ ] 19.10.x (master)

<h2> How this pull request can be tested ? </h2>

Main menu on the left

### CODE REVIEW

File: www/front_src/src/components/navigation/index-old.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
+import React, { Component } from "react";
```
Comment (bdauria) [inline www/front_src/src/components/navigation/index-old.js]: Is this file still needed? otherwise I think we can get rid of it

Comment (victorvassilev) [reply www/front_src/src/components/navigation/index-old.js]: I have absolutely no idea :) @danielhalupka ?

Comment (victorvassilev) [reply www/front_src/src/components/navigation/index-old.js]: I removed it

**Comment by vhr** [comment]
> Please to respect the PHP coding style => camel case

If you mean Entity classes they represent DB table columns structure, and we fetch result from DB as an object, not array for that they have to be with the same name

`$stmt->setFetchMode(PDO::FETCH_CLASS,` Topology::class);`

**Comment by sc979** [comment]
please correct the specified coding style listed on jenkins :
image

File: src/Centreon/Application/DataRepresenter/Topology/NavigationList.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+namespace Centreon\Application\DataRepresenter\Topology;
+
+use JsonSerializable;
+
+class NavigationList implements JsonSerializable
+{
+
+    /**
+     * @var array
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/NavigationList.php]: what is it used for ?

File: src/Centreon/Application/DataRepresenter/Topology/NavigationList.php [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                    'is_react' => (bool)$entity->getIsReact()
+                ];
+            } elseif (preg_match('/^(\d)(\d\d)(\d\d)$/', $entity->getTopologyPage(), $matches)) { // level 3
+                $levelTwo = $matches[1] . $matches[2];
+
+                //level 3 items can be grouped for better display
+
+                //make sure we skip groups (we extracted them above)
+                if (!(is_null($entity->getTopologyPage()) && $entity->getIsReact() == '0')) {
+
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/NavigationList.php]: sonar will ask to remove this whitespace

File: src/Centreon/Application/DataRepresenter/Topology/NavigationList.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                        'page' => $entity->getTopologyPage(),
+                        'label' => $entity->getTopologyName(),
+                        'url' => $entity->getTopologyUrl(),
+                        'options' => $entity->getTopologyUrlOpt(),
+                        'is_react' => (bool)$entity->getIsReact()
+                    ];
+
+                    //check if topology has group index
+                    if (!is_null($entity->getTopologyGroup()) && isset($groups[$levelTwo][$entity->getTopologyGroup()])) {
+
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/NavigationList.php]: sonar will ask to remove this whitespace

File: src/Centreon/Application/DataRepresenter/Topology/NavigationList.php [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    /**
+     * Extract the array without keys to avoid serialization into objects
+     * @param $naviList
+     * @return array
+     */
+    private function removeKeysFromArray($naviList)
+    {
+        foreach ($naviList as $key => &$value) {
+
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/NavigationList.php]: sonar will ask to remove this whitespace

File: src/Centreon/Application/DataRepresenter/Topology/NavigationList.php [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            if (!empty($value['children'])) {
+                foreach ($value['children'] as $k => &$c) {
+                    if (!empty($c['groups'])) {
+                        $c['groups'] = array_values($c['groups']);
+                    }
+                }
+                $value['children'] = array_values($value['children']);
+            }
+
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/NavigationList.php]: sonar will ask to remove this whitespace

File: src/Centreon/Application/DataRepresenter/Topology/ReactAcl.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/ReactAcl.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

**Comment by sc979** [comment]
I thought that the license should be apache like instead of GPL, wasn't it ?

File: src/Centreon/Application/DataRepresenter/Topology/ReactAcl.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace Centreon\Application\DataRepresenter\Topology;
+
+use JsonSerializable;
+
+class ReactAcl implements JsonSerializable
+{
+
+    /**
+     * @var array
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/ReactAcl.php]: comment on what it is used for

File: src/Centreon/Application/DataRepresenter/Topology/ReactAclForActive.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/Centreon/Application/DataRepresenter/Topology/ReactAclForActive.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 10]
```diff
@@ -75,6 +93,82 @@ public function getGetTopologyByPage(): array
...
+     *   )
+     * )
+     * @throws \RestBadRequestException
+     */
+    public function getMenuList()
+    {
+        $user = $this->getDi()[ServiceProvider::CENTREON_USER];
+
+        if (empty($user)) {
+            throw new \RestBadRequestException('User not found in session. Please relog.');
```
Comment (sc979) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]: shouldn't we add a log in login.log using centreonLog object -> to help support in their tasks ?

File: src/Centreon/Domain/Repository/TopologyRepository.php [hunk 11]
```diff
@@ -36,6 +40,7 @@ public function enableMenus(): bool
 
     /**
      * Get Topologies according to ACL for user
+     * @todo refactor this into function below it
      */
```
Comment (sc979) [inline src/Centreon/Domain/Repository/TopologyRepository.php]: @param -> what type is $user ? a string ? an array ?

File: src/Centreon/Domain/Repository/TopologyRepository.php [hunk 12]
```diff
@@ -72,7 +77,7 @@ public function getReactTopologiesPerUserWithAcl($user)
                             . "AND acl_topology_relations.acl_topo_id = '" . $topo_group["acl_topology_id"] . "' ";
                         $DBRESULT2 = $this->db->query($query2);
                         while ($topo_page = $DBRESULT2->fetchRow()) {
-                            $topology[] = (int) $topo_page["topology_topology_id"];
+                            $topology[] = (int)$topo_page["topology_topology_id"];
```
Comment (sc979) [inline src/Centreon/Domain/Repository/TopologyRepository.php]: line 72 + 79 + others -> replace fetchRow with fetch

File: src/Centreon/Domain/Repository/TopologyRepository.php (hunk 11 above)
Comment (sc979) [reply src/Centreon/Domain/Repository/TopologyRepository.php]: @return -> what type is $topologyUrls

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    protected $params;
+
+    /**
+     * Add a column
+     *
+     * @param string $parameter
+     * @param mixed $value
+     * @param int $data_type
+     */
+    public function addColumn($parameter, $value, int $data_type = PDO::PARAM_STR)
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
    public function addColumn(string $parameter, $value, int $data_type = PDO::PARAM_STR)
```

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Add a value
+     *
+     * @param string $parameter
+     * @param mixed $value
+     * @param int $data_type
+     */
+    public function addValue($parameter, $value, int $data_type = PDO::PARAM_STR)
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
    public function addValue(string $parameter, $value, int $data_type = PDO::PARAM_STR)
```

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Add a parameter
+     *
+     * @param string $parameter
+     * @param mixed $value
+     * @param int $data_type
+     */
+    public function addParam($parameter, $value, int $data_type = PDO::PARAM_STR)
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
    public function addParam(string $parameter, $value, int $data_type = PDO::PARAM_STR)
```

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Bind collected data to statement
+     *
+     * @param PDOStatement $stmt
+     */
+    public function bind(PDOStatement $stmt)
+    {
+        // bind columns to statment
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
        // bind columns to statement
```

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        foreach ($this->values as $parameter => $data) {
+            $stmt->bindColumn($parameter, $data['value'], $data['data_type']);
+        }
+
+        // bind values to statment
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
        // bind values to statement
```

File: src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        foreach ($this->values as $parameter => $data) {
+            $stmt->bindValue($parameter, $data['value'], $data['data_type']);
+        }
+
+        // bind parameters to statment
```
Comment (sc979) [inline src/Centreon/Infrastructure/CentreonLegacyDB/StatementCollector.php]:
```suggestion
        // bind parameters to statement
```

File: src/Centreon/ServiceProvider.php [hunk 19]
```diff
@@ -75,6 +85,12 @@ class ServiceProvider implements AutoloadServiceProviderInterface
      */
```
Comment (sc979) [inline src/Centreon/ServiceProvider.php]: comment on the return type

**Comment by sc979** [comment]
And please correct every coding style linked to his PR and detailled in the jenkins job :
image

**Comment by callapa** [comment]
> If you mean Entity classes they represent DB table columns structure, and we fetch result from DB as an object, not array for that they have to be with the same name

You can create an entity creator that will use the data to correctly fill the entity (see example in the sf4-poc)

**Comment by victorvassilev** [comment]
all good for everyone?

**Pull request merged**